### PR TITLE
Limit link def expansion

### LINKED ISSUE #844: pathologic input: O(n^2) with link ref. defs
Input generated with this leads to O(n^2) both in time and output size:
```
$ python -c 'N=1000; print("[x]: " + "x" * N + "\n[x]" * N)'
```

(same as https://github.com/mity/md4c/issues/238; cmark seems to limit number of link. ref. defs instantiations in some way so it is not vulnerable.)

### PR DESCRIPTION
**Pull request by notriddle**
Fixes #844

### Before

image

### After

image

### CODE REVIEW

File: pulldown-cmark/src/parse.rs [hunk 1]
```diff
@@ -225,6 +243,8 @@ impl<'input, F: BrokenLinkCallback<'input>> Parser<'input, F> {
             inline_stack,
             link_stack,
             html_scan_guard,
+            // always allow 100KiB
+            link_ref_expansion_limit: text.len().max(100_000),
```
Comment (ollpu) [inline pulldown-cmark/src/parse.rs:247]: I guess I'd rather make this `text.len() + 100_000` but replicating cmark is good.

**Review by ollpu — DISMISSED** [review]
Simple enough solution. Looks good!

Similar quadratic growth could happen with the broken link callback, although it's highly context dependent. Should we warn about that in the docs?

**Comment by ollpu** [comment]
> This runs after the callback, so it should cover it, right?

Ah, I see. But running the callback already potentially involves generating long strings, so quadratic work could be done anyway.

I guess the same applies for ref defs: If the url/link are not borrowed from the input, then them being fetched and cloned from the map is already quadratic. An input like

```python
"[x]: " + "x\\." * N + "\n[x]" * N
```

is still quadratic (though only noticeably so with N > 20000 as it's just string cloning).

So the check for whether the limit has been exhausted should be before anything is cloned or the callback is called.

I'm also not sure that we should apply the same limit to the broken link callback. For ref defs, the url & title are included in the document itself once, so the limit will never be hit if each definition is only used once. This is not the case for links expanded by the callback. Of course the 100K minimum size will be hard to reach in practice.

**Comment by ollpu** [comment]
> That would be solved by making `CowStr::Boxed` use an `Arc` instead of a `Box`, right?

Yeah, that could be considered. For this issue though, there's probably a reasonable solution that checks the limit before cloning? Not sure if it's worth blocking this on switching to `Arc`s everywhere.

**Review by ollpu — APPROVED** [review]
Looks good now! (Assuming you're ok with the callback using the same limit)

**Comment by Martin1887** [comment]
I have to deep a bit more on this but it looks good, thanks!

**Pull request merged**